src/test/java/com/atlassian/jira/cloud/jenkins/provider/HttpClientProviderTest.java

```diff
@@ -0,0 +1,83 @@
+package com.atlassian.jira.cloud.jenkins.provider;
+
+import com.atlassian.jira.cloud.jenkins.BaseMockServerTest;
+import com.atlassian.jira.cloud.jenkins.HttpClientProviderTestGenerator;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.inject.Inject;
+import java.io.IOException;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class HttpClientProviderTest extends BaseMockServerTest {
+
+    @Inject
+    private OkHttpClient httpClient;
+
+    @Before
+    public void setup() throws IOException {
+        super.setup();
+        httpClient = new HttpClientProvider().httpClient();
+    }
+
+    @Test
+    public void testNoRetryFor2XX() throws IOException {
+        // setup
+        HttpClientProviderTestGenerator.succeedWith2XXOnInitialAttempt(this);
+        final Request request = getRequest();
+
+        // execute
+        final Response response = httpClient.newCall(request).execute();
+
+        // verify
+        assertThat(response.code()).isEqualTo(202);
+    }
+
+    @Test
+    public void testNoRetryFor4XX() throws IOException {
+        // setup
+        HttpClientProviderTestGenerator.failWith4XXOnInitialAttempt(this);
+        final Request request = getRequest();
+
+        // execute
+        final Response response = httpClient.newCall(request).execute();
+
+        // verify
+        assertThat(response.code()).isEqualTo(404);
+    }
+
+    @Test
+    public void testRetryOnceFor5XXAndSucceed() throws IOException {
+        // setup
+        HttpClientProviderTestGenerator.failWith503AndThenSucceed2XX(this);
+        final Request request = getRequest();
+
+        // execute
+        final Response response = httpClient.newCall(request).execute();
+
+        // verify
+        assertThat(response.code()).isEqualTo(202);
```

Comment:
/nit is there any way to check how many attempts were done?

Reply:
added extra assertions + comments now.